jujugui/static/gui/src/app/components/header-search/header-search.js

```diff
@@ -0,0 +1,187 @@
+
+      @method _generateClasses
+      @returns {String} The collection of class names.
+    */
+    _generateClasses: function() {
+      return classNames(
+        'header-search',
+        this.state.active ? 'header-search--active' : ''
+      );
+    },
+
+    /**
+      Generate the close button classes based on the props.
+
+      @method _closeClasses
+      @returns {String} The collection of class names.
+    */
+    _closeClasses: function() {
+      return classNames(
+        'header-search__close',
+        {
+          hidden: !this.state.active
+        }
+      );
+    },
+
+    /**
+      Handle the search input receiving focus.
+
+      @method _handleSearchFocus
+    */
+    _handleSearchFocus: function() {
+      this.setState({
+        active: true,
+        inputStyles: this._generateInputStyles(true)
+      });
+    },
+
+    /**
+      Handle the search input losing focus.
+
+      @method _handleSearchBlur
+    */
+    _handleSearchBlur: function() {
+      if (!this.state.search && !this.state.query) {
+        this.setState({
+          active: false,
+          inputStyles: this._generateInputStyles(false)
+        });
+      }
+    },
+
+    /**
+      Set the width of the input based on the window size.
+
+      @method _generateInputStyles
+      @param {Boolean} active Set the width only if it active
+      @returns {Object} The object of styles
+    */
+    _generateInputStyles: function(active) {
+      var styles = {};
+      if (active) {
+        styles.width = (window.innerWidth * 0.40) + 'px';
+      }
+      return styles;
+    },
+
+    /**
+      Make search when the search form is submitted.
+
+      @method _handleSubmit
+      @param {Object} e The submit event
+    */
+    _handleSubmit: function(e) {
+      e.preventDefault();
+      this.props.changeState({
+        sectionC: {
+          metadata: {
+            search: {
+              text: this.state.query
+            }
+          }
+        }
+      });
+    },
+
+    /**
+      Close the header and sectionC when the button is clicked.
+
+      @method _handleClose
+    */
+    _handleClose: function() {
+      this.props.changeState({
+        sectionC: {
+          component: null,
+          metadata: {}
+        }
+      });
+    },
+
+    render: function() {
+      return (
+        <div className={this._generateClasses()}>
+          <form className="header-search__form">
+            <button type="submit"
+              onClick={this._handleSubmit}
+              className="header-search__submit"
+              dangerouslySetInnerHTML={{__html: this.icons.submit}} />
+            <input type="search" name="query"
+              className="header-search__input"
+              placeholder="Search the store"
+              onFocus={this._handleSearchFocus}
+              onBlur={this._handleSearchBlur}
+              style={this.state.inputStyles}
+              valueLink={this.linkState('query')} />
```

Comment:
Is this really any better? I'm a little worried those who aren't in the know will not understand what the mixin does.